src/index.js

```diff
@@ -1,117 +1,93 @@
+const isObj = o => typeof o === 'object'
+
+const each = (obj, fun) => {
+  if ('forEach' in obj) obj.forEach(fun)
+  else {
+    const iterator = isObj(obj) ? Object.entries(obj) : obj
+    for (const [key, val] of iterator) fun(val, key, obj)
+  }
+  return obj
+}
+
 const computedStack = []
 const observersMap = new WeakMap()
 
-const computed = function(fun, { autoRun = true, callback = null } = {}) {
-    const proxy = new Proxy(fun, {
-        apply: function(target, thisArg, argsList) {
-            const performComputation = (fun = null) => {
-                computedStack.unshift(callback || proxy)
-                const result = fun ? fun() : target.apply(thisArg, argsList)
-                computedStack.shift()
-                return result
-            }
-
-            argsList.push({
-                computeAsync: target => performComputation(target)
-            })
-
-            return performComputation()
-        }
-    })
-    if(autoRun) { proxy() }
-    return proxy
+const computed = (fun, {autoRun = true, callback = null} = {}) => {
+  const proxy = new Proxy(fun, {
+    apply (target, context, args) {
+      const performComputation = (fun = null) => {
+        computedStack.unshift(callback || proxy)
+        const result = fun ? fun() : target.apply(context, args)
+        computedStack.shift()
+        return result
+      }
+
+      args.push({computeAsync: performComputation})
+      return performComputation()
+    }
+  })
+  if (autoRun) proxy()
+  return proxy
 }
 
-const dispose = _ => _.__disposed = true
+const dispose = o => { o.__disposed = true }
 
-const batcher = {
-    timeout: null,
-    queue: new Set(),
-    process: () => {
-        for(let task of batcher.queue)
-            task()
-        batcher.queue.clear()
-        batcher.timeout = null
-    },
-    enqueue: task => {
-        if(batcher.timeout === null)
-            batcher.timeout = setTimeout(batcher.process, 0)
-        batcher.queue.add(task)
-    }
+const runAsync = async f => (await new Promise(resolve => resolve(f)))()
+
+const batcher = {
+  queue: new Set(),
+  process () {
+    each(batcher.queue, task => task()).clear()
+  },
+  enqueue (task) {
+    runAsync(batcher.process)
```

Comment:
With the previous implementation (setTimeout) I could prevent more than one execution by checking if there was an active batch.

The issue here is that now `enqeue` creates a new Promise with every call, meaning that `process` will run multiple times.

Reply:
🤔 hmm yeah, you're right I din't think that far. Oh well now I know 👍